observability/prometheus-backend/src/main/scala/sttp/client4/prometheus/PrometheusBackend.scala

```diff
@@ -132,19 +138,25 @@ object PrometheusBackend {
   Hence, we need to store a global cache o created histograms/gauges, so that we can properly re-use them.
    */
 
-  private val histograms = new mutable.WeakHashMap[CollectorRegistry, ConcurrentHashMap[String, Histogram]]
-  private val gauges = new mutable.WeakHashMap[CollectorRegistry, ConcurrentHashMap[String, Gauge]]
-  private val counters = new mutable.WeakHashMap[CollectorRegistry, ConcurrentHashMap[String, Counter]]
-  private val summaries = new mutable.WeakHashMap[CollectorRegistry, ConcurrentHashMap[String, Summary]]
+  private def unregister[T <: Collector](prometheusRegistry: PrometheusRegistry, collectors: mutable.WeakHashMap[PrometheusRegistry, ConcurrentHashMap[String, T]]): Unit =
+    collectors
+      .getOrElse(prometheusRegistry, new ConcurrentHashMap[String, T]())
+      .values()
+      .forEach(c => prometheusRegistry.unregister(c))
+
+  private val histograms = new mutable.WeakHashMap[PrometheusRegistry, ConcurrentHashMap[String, Histogram]]
+  private val gauges = new mutable.WeakHashMap[PrometheusRegistry, ConcurrentHashMap[String, Gauge]]
+  private val counters = new mutable.WeakHashMap[PrometheusRegistry, ConcurrentHashMap[String, Counter]]
+  private val summaries = new mutable.WeakHashMap[PrometheusRegistry, ConcurrentHashMap[String, Summary]]
 
   private def cacheFor[T](
-      cache: mutable.WeakHashMap[CollectorRegistry, ConcurrentHashMap[String, T]],
-      collectorRegistry: CollectorRegistry
+      cache: mutable.WeakHashMap[PrometheusRegistry, ConcurrentHashMap[String, T]],
+      prometheusRegistry: PrometheusRegistry
   ): ConcurrentHashMap[String, T] =
     cache.synchronized {
```

Comment:
as a follow-up task, we should remove the `.synchronized`, as it shouldn't be used in Java21+, and repalce it with some kind of a concurrent data structure (or improve the cache in general)

Reply:
or maybe we can do it right away, this is invoked only during the backend's construction, so a simple lock should suffice in fact; but we should also protect the cleanup method, then

Reply:
how about separate PR focused on cache I would prefer not to mix it with this update

Reply:
ok :)